### ISSUE TITLE
Chrome back-end - Devtools API/Puppeteer?

### ISSUE THREAD

**Issue by barryvdh** [KnpLabs/snappy#284]
I see you just merged Chrome back-end, which is awesome. But as far as I can tell, the commandline options are pretty limited (eg. no page size etc). The devtools options are more extensive: https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF

Google also released Puppeteer, which is a Nodejs api to render using the Devtools API, so that might also be interesting to use: https://github.com/GoogleChrome/puppeteer
A PHP examle can also be seen here: https://github.com/spatie/browsershot/pull/85

Or using the devtools socket directly: https://gitlab.com/DMore/chrome-mink-driver/

**Comment by NiR-** [comment]
Indeed chrome-cli is pretty limited, especially for screenshots. I wanted to give it a try, but finally I'm not sure if it's the right move.

I think the `chrome-mink-driver` would be a right solution, as it doesn't imply dependency on another language and limit overhead due to firing up multiple binaries. I also realise that it's possible to use it across a network (useful for containerized envs).

I'll give it a try and we would be able to compare both PRs and see their pros/cons :)

**Comment by barryvdh** [comment]
Yeah with the mink driver it would be also interesting to have a long running chrome instance perhaps, instead of fire and kill every time (for a specific use case)

**Comment by barryvdh** [comment]
I've been testing this in a few project, and I'm actually quite happy with Puppeteer so far. It's easy to setup (just install it with NPM, which installs Chrome also) and reliable. It does some async/await and socket stuff etc, so it probably isn't as easy/reliable to do that in PHP.

From what I can tell it's a lot better then wkhtmltopdf for certain cases.

**Comment by NiR-** [comment]
Hey @barryvdh,

Thank you for this precious feedback! I'm going to not work on sf/php projects in upcoming months, so it's really valuable to have feedback based on real projects (rather than symfony-demo and other toy projets) :)

As already said, chrome-cli is very limited whereas puppeteer does a better job with more possibilities. I'm not really happy with the idea of a hard requirement on a NodeJS package, but as it's a sort of DX tradeoff I find it acceptable. So I'll eventually remove chrome-cli in favor of puppeteer ;) However, I think we'd still benefit from a proper PHP implementation (so it'd be possible to use chrome as a backend without the hard requirement and with more flexibility, I guess).
